File: 04-implementation/backend/app/models/analysis.py
from datetime import datetime
from sqlalchemy import String, Integer, Float, Text, ForeignKey, DateTime, JSON
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class GapAnalysis(Base):
    __tablename__ = "gap_analyses"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    programme_id: Mapped[int] = mapped_column(ForeignKey("programmes.id"), nullable=False)
    # For programme-to-market: career_path is set; compare_programme_id is null
    # For programme-to-programme: compare_programme_id is set; career_path is null
    career_path: Mapped[str] = mapped_column(String(64), nullable=True)
    compare_programme_id: Mapped[int] = mapped_column(ForeignKey("programmes.id"), nullable=True)
    scenario: Mapped[str] = mapped_column(String(64), default="core")  # core | core_electives | hypothetical
    status: Mapped[str] = mapped_column(String(50), default="pending")
    celery_task_id: Mapped[str] = mapped_column(String(255), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    completed_at: Mapped[datetime] = mapped_column(DateTime, nullable=True)

    result: Mapped["GapResult"] = relationship(back_populates="analysis", uselist=False, cascade="all, delete-orphan")


class GapResult(Base):
    __tablename__ = "gap_results"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    analysis_id: Mapped[int] = mapped_column(ForeignKey("gap_analyses.id"), nullable=False, unique=True)
    kl_divergence: Mapped[float] = mapped_column(Float, nullable=True)
    cosine_similarity: Mapped[float] = mapped_column(Float, nullable=True)
    # JSON: list of {skill, gap_score, rca_weight, direction: "deficit"|"surplus"|"common"}
    ranked_gaps: Mapped[dict] = mapped_column(JSON, nullable=True)
    # JSON: {common: [...], programme_unique: [...], market_unique: [...]}
    skill_decomposition: Mapped[dict] = mapped_column(JSON, nullable=True)
    # JSON: courses × skills matrix for heatmap
    heatmap_data: Mapped[dict] = mapped_column(JSON, nullable=True)
    narrative_summary: Mapped[str] = mapped_column(Text, nullable=True)
    pdf_path: Mapped[str] = mapped_column(String(512), nullable=True)

    analysis: Mapped["GapAnalysis"] = relationship(back_populates="result")

File: 04-implementation/backend-rust/shared/src/models/analysis.rs
use chrono::NaiveDateTime;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::FromRow;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct GapAnalysis {
    pub id: i32,
    pub programme_id: i32,
    pub career_path: Option<String>,
    pub compare_programme_id: Option<i32>,
    pub scenario: String, // "core" | "core_electives" | "hypothetical"
    pub status: String,   // "pending" | "running" | "completed" | "failed"
    pub celery_task_id: Option<String>,
    pub created_at: NaiveDateTime,
    pub completed_at: Option<NaiveDateTime>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct GapResult {
    pub id: i32,
    pub analysis_id: i32,
    pub kl_divergence: Option<f64>,
    pub cosine_similarity: Option<f64>,
    pub ranked_gaps: Option<Value>,
    pub skill_decomposition: Option<Value>,
    pub heatmap_data: Option<Value>,
    pub narrative_summary: Option<String>,
    pub pdf_path: Option<String>,
}
